mono/sgen/sgen-marksweep.c

```diff
@@ -2039,6 +2060,9 @@ major_start_major_collection (void)
 		sgen_evacuation_freelist_blocks (&free_block_lists [MS_BLOCK_FLAG_REFS][i], i);
 	}
 
+	/* We expect workers to have very few blocks on the freelist, just evacuate them */
+	sgen_workers_foreach (sgen_worker_clear_free_block_lists_evac);
```

Comment:
The problem arises from workers' lists having free blocks at the start of a major (which was not the case before since worker free block lists were not used before having parallel nurseries, they are cleared after a major finish). During evacuation collections, block that are to be evacuated must not reside on any free block lists, since we only evacuate and not allocate from them.